File: src/main/java/org/kjots/json/object/simple/impl/SimpleJsonObjectFactoryImpl.java
/* 
 * Copyright © 2010 Karl J. Ots <kjots@kjots.org>
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.kjots.json.object.simple.impl;

import java.lang.reflect.InvocationTargetException;

import org.kjots.json.object.shared.JsonArray;
import org.kjots.json.object.shared.JsonBooleanArray;
import org.kjots.json.object.shared.JsonBooleanMap;
import org.kjots.json.object.shared.JsonNumberArray;
import org.kjots.json.object.shared.JsonNumberMap;
import org.kjots.json.object.shared.JsonObject;
import org.kjots.json.object.shared.JsonObjectArray;
import org.kjots.json.object.shared.JsonObjectFactory;
import org.kjots.json.object.shared.JsonObjectMap;
import org.kjots.json.object.shared.JsonStringArray;
import org.kjots.json.object.shared.JsonStringMap;
import org.kjots.json.object.simple.SimpleJsonObjectGenerator;
import org.kjots.json.object.simple.SimpleJsonValue;

import com.google.inject.Inject;

/**
 * Simple JSON Object Factory Implementation.
 * <p>
 * Created: 5th March 2010
 *
 * @author <a href="mailto:kjots@kjots.org">Karl J. Ots &lt;kjots@kjots.org&gt;</a>
 */
public class SimpleJsonObjectFactoryImpl extends JsonObjectFactory {
  /** The JSON object generator. */
  @Inject
  private SimpleJsonObjectGenerator jsonObjectGenerator;
  
  /**
   * Create a new JSON object with the given underlying JSON object.
   *
   * @param <T> The type of the JSON object.
   * @param jsonObjectClass The class of the JSON object.
   * @param object The underlying JSON object.
   * @return The JSON object.
   */
  @Override
  @SuppressWarnings("unchecked")
  public <T extends JsonObject> T createJsonObject(Class<T> jsonObjectClass, Object object) {
    SimpleJsonObjectImpl jsonObjectImpl = this.createStaticJsonObject(jsonObjectClass, (SimpleJsonValue)object);
    if (jsonObjectImpl == null) {
      Class<? extends SimpleJsonObjectImpl> jsonObjectImplClass = this.jsonObjectGenerator.getJsonObjectImplClass(jsonObjectClass);
      
      try {
        jsonObjectImpl = jsonObjectImplClass.getConstructor(SimpleJsonValue.class).newInstance(object);
      }
      catch (NoSuchMethodException nsme) {
        throw new IllegalStateException(nsme);
      }
      catch (IllegalAccessException iae) {
        throw new IllegalStateException(iae);
      }
      catch (InstantiationException ie) {
        throw new IllegalStateException(ie);
      }
      catch (InvocationTargetException ite) {
        Throwable t = ite.getCause();
        
        if (t instanceof RuntimeException) {
          throw (RuntimeException)t;
        }
        else if (t instanceof Error) {
          throw (Error)t;
        }
        else {
          throw new IllegalStateException(t);
        }
      }
    }
    
    return (T)jsonObjectImpl;
  }

  /**
   * Create a new JSON object.
   *
   * @param <T> The type of the JSON object.
   * @param jsonObjectClass The class of the JSON object.
   * @return The JSON object.
   */
  @Override
  public <T extends JsonObject> T createJsonObject(Class<T> jsonObjectClass) {
    return this.createJsonObject(jsonObjectClass, SimpleJsonValue.createObject());
  }

  /**
   * Create a new JSON array.
   *
   * @param <T> The type of the JSON array.
   * @param jsonArrayClass The class of the JSON array.
   * @return The JSON array.
   */
  @Override
  public <T extends JsonArray> T createJsonArray(Class<T> jsonArrayClass) {
    return this.createJsonObject(jsonArrayClass, SimpleJsonValue.createArray());
  }
  
  
  /**
   * Create a new JSON object instance with given class using the given
   * underlying JSON object.
   * <p>
   * This method will only create JSON object instances with statically defined
   * implementations.
   *
   * @param jsonObjectClass The class of the JSON object.
   * @param object The underlying JSON object.
   * @return The JSON object.
   */
  private SimpleJsonObjectImpl createStaticJsonObject(Class<? extends JsonObject> jsonObjectClass, SimpleJsonValue object) {
    if (jsonObjectClass.equals(JsonObject.class)) {
      return new SimpleJsonObjectImpl(object);
    }
    else if (jsonObjectClass.equals(JsonArray.class)) {
      return new SimpleJsonArrayImpl(object);
    }
    else if (jsonObjectClass.equals(JsonBooleanArray.class)) {
      return new SimpleJsonBooleanArrayImpl(object);
    }
    else if (jsonObjectClass.equals(JsonNumberArray.class)) {
      return new SimpleJsonNumberArrayImpl(object);
    }
    else if (jsonObjectClass.equals(JsonStringArray.class)) {
      return new SimpleJsonStringArrayImpl(object);
    }
    else if (jsonObjectClass.equals(JsonObjectArray.class)) {
      return new SimpleJsonObjectArrayImpl<JsonObject>(object);
    }
    else if (jsonObjectClass.equals(JsonBooleanMap.class)) {
      return new SimpleJsonBooleanMapImpl(object);
    }
    else if (jsonObjectClass.equals(JsonNumberMap.class)) {
      return new SimpleJsonNumberMapImpl(object);
    }
    else if (jsonObjectClass.equals(JsonStringMap.class)) {
      return new SimpleJsonStringMapImpl(object);
    }
    else if (jsonObjectClass.equals(JsonObjectMap.class)) {
      return new SimpleJsonObjectMapImpl<JsonObject>(object);
    }
    else {
      return null;
    }
  }
}

File: org.kjots.json-toolkit.json-object.scala/src/main/scala/org/kjots/json/object/scala/impl/SimpleJsonObjectFactoryImpl.scala
/* 
 * Copyright © 2011 Karl J. Ots <kjots@kjots.org>
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *     http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.kjots.json.`object`.scala.impl

import org.kjots.json.`object`.impl.JvmJsonObjectFactoryImplBase
import org.kjots.json.`object`.scala.SimpleJsonObjectGenerator
import org.kjots.json.`object`.shared._
import org.kjots.json.`object`.simple.impl._
import org.kjots.json.`object`.simple.SimpleJsonValue

/**
 * Simple JSON Object Factory Implementation.
 * <p>
 * Created: 28th May 2011.
 *
 * @author <a href="mailto:kjots@kjots.org">Karl J. Ots &lt;kjots@kjots.org&gt;</a>
 * @since 1.2
 */
class SimpleJsonObjectFactoryImpl extends JvmJsonObjectFactoryImplBase {
  private lazy val JSON_OBJECT_CLASS = classOf[JsonObject]
  private lazy val JSON_ARRAY_CLASS = classOf[JsonArray]
  private lazy val JSON_BOOLEAN_ARRAY_CLASS = classOf[JsonBooleanArray]
  private lazy val JSON_NUMBER_ARRAY_CLASS = classOf[JsonNumberArray]
  private lazy val JSON_STRING_ARRAY_CLASS = classOf[JsonStringArray]
  private lazy val JSON_OBJECT_ARRAY_CLASS = classOf[JsonObjectArray[_]]
  private lazy val JSON_BOOLEAN_MAP_CLASS = classOf[JsonBooleanMap]
  private lazy val JSON_NUMBER_MAP_CLASS = classOf[JsonNumberMap]
  private lazy val JSON_STRING_MAP_CLASS = classOf[JsonStringMap]
  private lazy val JSON_OBJECT_MAP_CLASS = classOf[JsonObjectMap[_]]
  
  override def createJsonObject[T <: JsonObject](jsonObjectClass: Class[T], obj: AnyRef): T = {
    def simpleJsonValue = obj.asInstanceOf[SimpleJsonValue]

    def simpleJsonObjectImpl = jsonObjectClass match {
      case `JSON_OBJECT_CLASS` => new SimpleJsonObjectImpl(simpleJsonValue)
      case `JSON_ARRAY_CLASS` => new SimpleJsonArrayImpl(simpleJsonValue)
      case `JSON_BOOLEAN_ARRAY_CLASS` => new SimpleJsonBooleanArrayImpl(simpleJsonValue)
      case `JSON_NUMBER_ARRAY_CLASS` => new SimpleJsonNumberArrayImpl(simpleJsonValue)
      case `JSON_STRING_ARRAY_CLASS` => new SimpleJsonStringArrayImpl(simpleJsonValue)
      case `JSON_OBJECT_ARRAY_CLASS` => new SimpleJsonObjectArrayImpl(simpleJsonValue)
      case `JSON_BOOLEAN_MAP_CLASS` => new SimpleJsonBooleanMapImpl(simpleJsonValue)
      case `JSON_NUMBER_MAP_CLASS` => new SimpleJsonNumberMapImpl(simpleJsonValue)
      case `JSON_STRING_MAP_CLASS` => new SimpleJsonStringMapImpl(simpleJsonValue)
      case `JSON_OBJECT_MAP_CLASS` => new SimpleJsonObjectMapImpl(simpleJsonValue)
      case _ => SimpleJsonObjectGenerator.newJsonObjectImpl(jsonObjectClass, simpleJsonValue)
    }

    simpleJsonObjectImpl.asInstanceOf[T]
  }

  override def createJsonObject[T <: JsonObject](jsonObjectClassName: String, obj: AnyRef): T =
    createJsonObject(Class.forName(jsonObjectClassName).asInstanceOf[Class[T]], obj)

  override def createJsonObject[T <: JsonObject](jsonObjectClass: Class[T]): T =
    createJsonObject(jsonObjectClass, SimpleJsonValue.createObject())

  override def createJsonObject[T <: JsonObject](jsonObjectClassName: String): T =
    createJsonObject(jsonObjectClassName, SimpleJsonValue.createObject())

  override def createJsonArray[T <: JsonArray](jsonArrayClass: Class[T]): T =
    createJsonObject(jsonArrayClass, SimpleJsonValue.createArray())

  override def createJsonArray[T <: JsonArray](jsonArrayClassName: String): T =
    createJsonObject(jsonArrayClassName, SimpleJsonValue.createArray())
}
